### PR TITLE
fix(api): atomic progress writes via merge_progress_data rpc

### PR DESCRIPTION
**Pull request by DysektAI**
## Root cause

The api-gateway progress write handlers did `GET user_progress -> mutate whole JSONB blob in JS -> PATCH whole blob back`. Two concurrent writers (website session, TarkovMonitor, rapid task/objective/level updates) could silently overwrite each other's changes (classic lost update). A write against a missing progress row also succeeded silently (PATCH matching 0 rows returns 204).

## Fix

- New Supabase RPC `merge_progress_data` (`supabase/migrations/20260705130000_add_merge_progress_rpc.sql`): merges only the supplied keys server-side under `SELECT ... FOR UPDATE` and returns the affected row count.
  - `taskCompletions`: shallow merge of only the changed entries
  - `taskObjectives`: per-key deep merge (`||` of the objective object)
  - `set`: top-level merge (level, `lastApiUpdate`)
  - `SECURITY DEFINER`, execute revoked from `PUBLIC`/`anon`/`authenticated`, granted to `service_role` only
- `handleUpdateLevel`, `handleUpdateObjective`, `handleUpdateTask`, `handleUpdateTasks` now send partial merges via the RPC. Task handlers snapshot/diff completions to send only changed entries.
- Row-count check: RPC returning `0` (or non-numeric) surfaces `Progress row not found for user` instead of silently no-op'ing.
- Regression tests: RPC-payload assertions, missing-row error, and a concurrency test proving two writers with stale reads no longer clobber each other's unrelated keys.

## Reviewer judgment call

`handleUpdateObjective` still GETs the row before merging so it can merge `complete`/`count`/`timestamp` into the existing objective. The write is atomic, but the read side can still race another objective writer; last objective-state write wins, which we believe is the desired semantic. Flagged with a code comment; pushing the field-level merge fully into SQL is a possible follow-up.

## Secondary: api.tarkovtracker.org host migration scaffolding

- `robots.txt` on the API host (crawlers off the API surface; docs/openapi stay public).
- `LEGACY_API_REDIRECT` env var (default `"false"`). When flipped to `"true"`, legacy `/api` and `/api/v2` requests on the main host return `308` to `api.tarkovtracker.org` with `Deprecation` and `Link: rel="successor-version"` headers, preserving path and query. Note (also in `docs/API.md`): .NET `HttpClient` and several other stacks drop `Authorization` on cross-host redirects, so clients should migrate proactively.

## Deploy notes

- **The SQL migration must be applied to Supabase before the worker ships** (`supabase db push` or manual), otherwise all progress writes fail with HTTP 404 on the RPC.
- After the migration is applied, `npm run supabase:types` can be re-run to pick up the new RPC in generated types (the gateway itself uses hand-written types and raw fetch, so this is not required for this PR).

## Test plan

- `npm run test:api-gateway` green (38 tests, incl. new missing-row, concurrency, and redirect on/off tests)
- `npm run typecheck` green (app + `workers/api-gateway/tsconfig.json`)
- `npm run lint` clean (0 errors, 0 warnings)
- Manual: a write against a missing progress row now returns a 500 with `Progress row not found for user`

## Summary by CodeRabbit

* **New Features**
  * Progress updates now merge more reliably, preserving existing data when only part of a task or level changes.
  * Legacy progress URLs can now be redirected to the new API host during migration.

* **Bug Fixes**
  * Reduced the chance of progress updates overwriting unrelated task data.
  * Improved handling of concurrent progress updates and missing progress records.

* **Documentation**
  * Added migration guidance and notes about the new progress API host and redirect behavior.

___

## **CodeAnt-AI Description**
**Save progress updates atomically and start the API host migration**

### What Changed
- Progress updates now merge only the changed parts, so concurrent task, objective, and level writes no longer overwrite each other
- Updates against a missing progress row now return an error instead of failing silently
- Task objective updates no longer reuse stale saved values when only the count or completion state changes
- The API host now serves `robots.txt` and can redirect legacy `/api` and `/api/v2` traffic to `api.tarkovtracker.org` with deprecation headers when enabled
- Tests now cover atomic merges, missing-row errors, and the legacy redirect behavior

### Impact
`✅ Fewer lost progress updates`
`✅ Clearer errors when a progress profile is missing`
`✅ Safer API host migration for authenticated clients`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

**Comment by DysektAI** [comment]
Addressed both review findings in 584ad3c0:

1. **CodeAnt (Major)** — `merge_progress_data` now rejects `NULL` `p_field` (`NOT IN` returns NULL for null operands, so the guard was skipped and writes silently fell through to `pve_data`). Added `p_field IS NULL OR` to the guard.

2. **CodeRabbit (Major)** — `handleUpdateObjective` no longer does a pre-merge GET. The merge payload is built from `update` only, so a count-only update can't carry back a stale `complete` (and vice versa). The RPC's per-key objective merge (`COALESCE(existing, '{}') || v_value`) preserves untouched fields server-side. Added a regression test verifying a count-only update sends no `complete` key.

Validation: `npm run test:api-gateway` 39/39, `npm run lint` clean, `npx tsc --noEmit -p workers/api-gateway/tsconfig.json` clean.

**Comment by DysektAI** [comment]
## Review pass 3 — findings fixed in e3f52faf

Traced the RPC SQL, all four handler write paths, the 308 redirect, and the tests. Validated the migration against a real Postgres 16 (behavioral tests below), not just the mocks. Three findings, all fixed:

**1. JSONB \`||\` corruption edge cases in \`merge_progress_data\` (medium)**
\`jsonb ||\` degrades to array concatenation when an operand isn't an object, and \`jsonb_set\` errors on scalar roots. A stored \`'null'::jsonb\` blob (JSON null, which \`COALESCE\` does *not* catch), a non-object \`taskCompletions\`/\`taskObjectives\`, or a non-object param could corrupt the whole blob into an array. Fixed: params are now validated with \`jsonb_typeof\` (raise on non-object), and the blob + nested containers are normalized to \`{}\` when non-object.

**2. Deploy-ordering hazard (ops)**
Workers auto-deploy from \`main\` via the Cloudflare Git integration, but migrations are applied manually *after* merge per the runbook. If this PR merges before \`supabase db push\`, **every** API progress write 404s on the missing RPC. Fixed: documented in the migration header and \`docs/runbook.md\` — **apply this migration to production before merging** (adding a function ahead of its caller is safe).

**3. \`Deprecation: true\` uses obsolete draft syntax (low)**
RFC 9745 (final) defines Deprecation as a structured-field Date. Now sends \`@1783296000\` (2026-07-06T00:00:00Z). Test updated.

**Verified against real Postgres 16:** combined completions/objectives/set merge preserves untouched keys; per-key objective deep-merge keeps \`complete\` when only \`count\` is sent; missing row returns 0; JSON-null and non-object blobs normalize instead of corrupting; array \`p_set\` and bad \`p_field\` raise. Concurrency proven with two overlapping transactions: writer B blocked on \`FOR UPDATE\` until A committed, then merged on top — both writes survived, unrelated keys intact.

**Also verified, no change needed:** \`SECURITY DEFINER\` + service_role-only grant + pinned \`search_path\` correct, params fully parameterized (no injection surface); redirect preserves path+query, skips CORS preflight, and only fires when the var is exactly \`"true"\`; task/batch handlers still read-then-cascade but only merge *changed* keys, so the residual race is last-writer-wins on the same key (inherent, acceptable); \`0\`-row RPC result fails loudly; docs match code.

**Validation:** \`test:api-gateway\` 39/39 ✓, \`tsc --noEmit\` (workers) ✓, \`lint\` 0 warnings ✓, \`supabase:check\` (full reset + lint) ✓.

**Verdict: production-ready**, with the one operational requirement above — push the migration before merging.

File: workers/api-gateway/src/__tests__/gateway.test.ts [hunk 1]
```diff
@@ -77,10 +115,13 @@ const createBaseFetchMock = ({
...
-      const patchBody = JSON.parse(String(init?.body || '{}')) as Record<string, unknown>;
-      onPatch?.(patchBody);
-      return new Response(null, { status: 204 });
+    if (url.includes('/rest/v1/rpc/merge_progress_data')) {
+      const payload = JSON.parse(String(init?.body || '{}')) as MergeRpcPayload;
+      onMerge?.(payload);
+      if (mergeStore) {
+        mergeStore.data = applyMergeRpc(mergeStore.data, payload);
+      }
+      return new Response(mergeResult ?? '1', { status: 200 });
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/__tests__/gateway.test.ts]: Fixed in c8263817 — the mock now only applies `applyMergeRpc` to `mergeStore` when the RPC result indicates an updated row (`Number(result) > 0`), matching production semantics where `mergeResult: '0'` is treated as a failed missing-row write.

**Comment by DysektAI** [comment]
## Final production-readiness pass — c8263817

**Review threads:** all resolved.
- CodeAnt (NULL `p_field`): fixed in 584ad3c0, thread resolved.
- CodeRabbit outside-diff (objective merge race): fixed in 584ad3c0 (patch built from `update` only).
- CodeRabbit inline / Kilo (test mock mutating `mergeStore` before checking `mergeResult`): **fixed in c8263817** — mock now only applies the merge when the RPC result reports an updated row.

**Also fixed in c8263817 — worker test typecheck gap:** `workers/api-gateway/tsconfig.json` excluded `src/**/__tests__/**`, so `tsc --noEmit` never covered the test file (hence IDE `DurableObjectState` errors and zero type safety in tests). Removed the exclusion; `npm run typecheck` in the worker (already wired into the CI `workers` job, which installs both root and worker deps) now typechecks the tests. `tsc --listFiles` confirms `gateway.test.ts` is included; zero errors.

**Rolling-deploy compatibility (verified):**
- Migration-first (documented order): safe — migration only adds `merge_progress_data` + grants; old worker never calls it.
- Worker-first (wrong order): loud failure, no silent path — missing RPC → PostgREST 404 → `res.ok` false → throw → 500 on progress writes. The updated-rows check also throws on `0`.
- `LEGACY_API_REDIRECT="false"` (set in wrangler.toml): redirect branch requires exactly `"true"`; legacy `/api/v2/*` routing is byte-identical until flipped.
- No request/response shape changes: progress GET path untouched; POST handlers return identical bodies to main.
- Env vars: only new read is `LEGACY_API_REDIRECT` (optional, set in `[vars]`).

**Validation:** `test:api-gateway` 39/39 ✅ · worker `tsc --noEmit` (incl. tests) ✅ · `lint` 0 warnings ✅ · Nuxt `typecheck` ✅ · `supabase:check` no schema errors ✅

**Go for merge** — prerequisite: `supabase db push --linked` the `20260705130000_add_merge_progress_rpc.sql` migration to production **before** merging.
